macros/shared/src/main/scala/zio/prelude/Assertion.scala

```diff
@@ -76,6 +76,10 @@ object Assertion {
 
   def matches(regexString: String): Assertion[String] = Matches(regexString)
 
+  def predicate[A](f: A => Boolean): Assertion[A] = Predicate(f, "custom function")
+
+  def predicate[A](f: A => Boolean, name: String): Assertion[A] = Predicate(f, name)
+
```

Comment:
The one issue I foresee (And, so so so so sorry for taking so damn long to respond 😅) is that the `Assertion` DSL was intentionally hamstrung in order to be completely "Unliftable" (able to be parsed and run at compile-time). I think we discussed once before that while this predicate type _can_ be read at compile-time with some trickery, that is only valid if the predicate is defined in a separate compilation cycle.

I previously added compile-errors to guide the user through this delicate process (in this file core/shared/src/main/scala-3/zio/prelude/Macros.scala), but if we add this `predicate` method, I think we'd at least need to further document that potential edge case. That's one option. Perhaps easier to implement, just some guidance in the case the user doesn't define their assertion in a separate compilation unit (which I fear might be the common case).

The other option would be to completely separate these two modes. Have your super-powered, super-flexible assertion _not_ use this DSL. And your validation type could accept a predicate directly. I wouldn't even bother with having an assertion DSL. In fact, I think that might be the better experience. The assertion DSL, while a step up from the type-level DSL (IMHO), is still not as nice as just writing normal scala functions—which your implementation allows for 😄 

I somewhat prefer the second approach. I think it's a bit cleaner, so when you're using one implementation, it isn't made slightly trickier or more awkward by having to deal with the concerns of the other.

I hope that all makes sense! And once again, apologies for the huge delay. Also, I might be not fully understanding the code, and completely open to discussion. Happy to do any pairing, or hop on a zoom call some time (maybe with our buddy @adamgfraser as well 👍 ) and helping to push this over the line. Thanks again for all your efforts,  Maxim. <3

Reply:
> The one issue I foresee (And, so so so so sorry for taking so damn long to respond 😅) is that the Assertion DSL was intentionally hamstrung in order to be completely "Unliftable" (able to be parsed and run at compile-time). I think we discussed once before that while this predicate type can be read at compile-time with some trickery, that is only valid if the predicate is defined in a separate compilation cycle.

> I previously added compile-errors to guide the user through this delicate process (in this file core/shared/src/main/scala-3/zio/prelude/Macros.scala), but if we add this predicate method, I think we'd at least need to further document that potential edge case. That's one option. Perhaps easier to implement, just some guidance in the case the user doesn't define their assertion in a separate compilation unit (which I fear might be the common case).

The way it works in this pr is that in dotty we would fail compilation, pointing out to the user that they need to switch to NewtypeCustom https://github.com/zio/zio-prelude/pull/793/files#diff-923a5a88adf9d9ce695dd7622dea3b6bb66a51ced1e3a88f939a7129087dea52R88. Potentially it makes sense to have more guidance for the user here, though I have no good ideas.

For scala 2 we can extract the function using eval, which seems to work even for functions defined in the same compilation unit. 
https://github.com/zio/zio-prelude/pull/793/files#diff-cb2c8d5aab7fb0cedcd2c4d8dbfad7777e0a9761ea3f7a5f09ad205695825880R206
Usage here:
https://github.com/zio/zio-prelude/pull/793/files#diff-109e476632c38a367dd0aa6740e67ad6801296fbed9592f638360a41d2334922R69
I'm not sure about the limitations of eval / I never managed to get it to blow up so far. Do we need additional safeguards here?

> The other option would be to completely separate these two modes. Have your super-powered, super-flexible assertion not use this DSL. And your validation type could accept a predicate directly. I wouldn't even bother with having an assertion DSL. In fact, I think that might be the better experience. The assertion DSL, while a step up from the type-level DSL (IMHO), is still not as nice as just writing normal scala functions—which your implementation allows for 😄

> I somewhat prefer the second approach. I think it's a bit cleaner, so when you're using one implementation, it isn't made slightly trickier or more awkward by having to deal with the concerns of the other.

An argument for keeping the assertion dsl is that the error messages in case of failed validation are much nicer this way.
It also allows users to combine the two, i.e. use normal assertions to catch 90% of invalid input with great error messages and use a lifted library function for the remaining ones. I could see something like this being useful for phone numbers or similar, though this is a bit of a contrived example.

Potentially it makes sense to rename `predicate` to `custom` to make it clearer that this is really an escape hatch. I'll do that and add a short docstring to it explaining the caveats.

> I hope that all makes sense! And once again, apologies for the huge delay. Also, I might be not fully understanding the code, and completely open to discussion. Happy to do any pairing, or hop on a zoom call some time (maybe with our buddy @adamgfraser as well 👍 ) and helping to push this over the line. Thanks again for all your efforts, Maxim. <3

Sounds great to me. Will ping you on discord 👍

Reply:
I played a bit around with it a bit more and I think you are right, having them be completely separate is the better way to go.
Updated the pr, please take another look